### DISCUSSION TITLE
Why does Connection.exec_driver_sql have inconsistent parameter parsing for queries with the IN operator?

### DISCUSSION THREAD

**Discussion by nicolaslegrand91** [sqlalchemy/sqlalchemy#7820] [Usage Questions] [answered]
These two queries are semantically identical, but one of them succeeds and the other one fails. The only difference is in the `WHERE` clause, where the two operands of the `OR` operator have been switched.

```python
from sqlalchemy import create_engine
engine = create_engine('mysql+pymysql://user:pwd@host:port/db', 

# Query succeeds (and does what's expected)
with engine.connect() as cn:
    cn.exec_driver_sql(
        f'UPDATE table SET column = "value" WHERE id = %s OR id IN %s', 
        (3, (1, 2))
    )

# Query fails
with engine.connect() as cn:
    cn.exec_driver_sql(
        f'UPDATE table SET column = "value" WHERE id IN %s OR id = %s', 
        ((1, 2), 3)
    )
```

Output of the failed query: 
> TypeError: 'int' object is not iterable

It seems that sqlalchemy argument parsing is dependant on the type of the first argument. If the first one if a tuple, the query fails, if it is a int/float/str, it succeeds. 

The workaround I've found so far is to use named arguments:

```python
# Query succeeds
with engine.connect() as cn:
    cn.exec_driver_sql(
        f'UPDATE table SET column = "value" WHERE id IN %(arg1)s OR id = %(arg2)s', 
        {'arg1': (1, 2), 'arg2': 3}
    )
```

However it is more verbose and I don't want to use this everywhere. Also note that PyMySQL cursor's `execute` method accepted both queries.

**Is there a reason for this behaviour?** 

_Note: This was a SO question, and the first reply mentions it could be interesting to have this posted here: https://stackoverflow.com/questions/71442220/why-does-connection-exec-driver-sql-have-inconsistent-parameter-parsing-for-quer_

**Comment by zzzeek** [comment, marked as answer]
I would classify this as a bug.     This issue only appears in version 1.4 and is not present in version 2.0, where parameter formats have been simplified.

To invoke this query use the cursor directly:

```
raw_connection = engine.raw_connection()
cursor = raw_connection.cursor()
cursor.execute("....", <params>)
```

**Reply by nicolaslegrand91** [reply]
Thanks for your reply. Do you want me to open an issue? Or do you know if 2.0 will be out shortly?

**Reply by zzzeek** [reply]
initial betas for 2.0 will be out in several months.

there's no plan to fix this for 1.4 right now as there's a workaround and I need all the time I can get to focus on 2.0.

A test for this case is proposed for 2.0 at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3705 so that this use case will continue to function.

**Reply by zzzeek** [reply]
the issue of parameter handling is very intricate, especially in the 1.x series, so fixing this for 1.4 would introduce a lot of regression risk.

**Reply by nicolaslegrand91** [reply]
Ok that's very clear, thanks a lot.
